### PR TITLE
Build an explicit abstract environment for check_regularity (fix #12334)

### LINKED ISSUE #12334: Erroneous "Constraints are not satisfied in this type" with mutually recursive abbreviations
When I say

```ocaml
type 'a t = 'a foo foo
and 'a foo = int constraint 'a = int
```

ocaml says

```
File "scratch.ml", line 1, characters 0-22:
1 | type 'a t = 'a foo foo
    ^^^^^^^^^^^^^^^^^^^^^^
Error: Constraints are not satisfied in this type.
       Type int should be an instance of 'a foo
       foo is abstract because no corresponding cmi file was found in path.
```

But this program should be accepted, I think, with something like this signature:

```ocaml
type 'a t = int constraint 'a = int
type 'a foo = int constraint 'a = int
```

I think the problem is that `Typedecl.check_regularity` uses the `orig_env` when checking arguments -- but the `orig_env` doesn't know that `'a foo` is an abbreviation for `int` when checking the manifest of `t`. I don't know enough of what's going on here to suggest a fix. But even if a proper fix is out of reach, the error message definitely has room for improvement: `int` *is* an instance of `'a foo` (because they're equal) and there is no missing cmi file involved.

### PR DESCRIPTION
**Pull request by garrigue**
The call to `check_regularity` in `Typecore` was using the original environment, which does not contain the newly added types. From a typing point of view, this is correct (those types will be handled as abstract), but this could cause strange error messages such as in #12334.
We now build an explicit abstract environment, extending `enter_type` with an extra switch.
Also fixed a bud in `check_regularity` which reported wrong constraints in a reversed order.

Note: it might be correct to used the partially abstract `temp_env` built above, but this requires a thorough examination.

### CODE REVIEW

File: typing/typedecl.ml [hunk 1]
```diff
@@ -1121,8 +1124,14 @@ let transl_type_decl env rec_flag sdecl_list =
     check_well_founded_decl new_env (List.assoc id id_loc_list) (Path.Pident id)
       decl to_check)
     decls;
+  (* Add abstract declarations to the environment to avoid strange error
+     messages (#12334) *)
+  let abs_env =
+    List.fold_left2 (enter_type ~abstract_abbrevs:false rec_flag)
+      env sdecl_list ids_list in
```
Comment (gasche) [inline typing/typedecl.ml:1143]: I would have expected `abstract_abbrevs:true` here. Are you using `true` anywhere?

File: testsuite/tests/typing-misc/constraints.ml [hunk 2]
```diff
@@ -433,3 +433,16 @@ and r = [s cstr t | `Bar]
...
+(* PR #12334 *)
+
+type 'a t = 'a foo foo
+and 'a foo = int constraint 'a = int
+[%%expect{|
+Line 1, characters 12-22:
+1 | type 'a t = 'a foo foo
+                ^^^^^^^^^^
+Error: Constraints are not satisfied in this type.
+       Type "'a foo foo" should be an instance of "int foo"
```
Comment (goldfirere) [inline testsuite/tests/typing-misc/constraints.ml]: It would be very nice to get an additional

```
Constraints cannot be inferred within a mutually recursive group of definitions
```

or something. Because `'a foo foo` *is* actually an instance of `int foo`.

Not sure how hard it would be to arrange for the better message.

File: typing/typedecl.ml (hunk 1 above)
Comment (goldfirere) [reply typing/typedecl.ml:1143]: Creating a whole new env here seems potentially expensive. Is there instead a way to catch and reinterpret the error message in `check_regularity`?

Comment (garrigue) [reply typing/typedecl.ml:1143]: @gasche Well-spotted. I did various testing, and using true or false doesn't seem to have much impact, but it should be `true` here for keeping the original behavior. Fixed now

@goldfirere Actually, this is not very expensive. We are just building a dummy env here, which contains only abstract types, and whose type parameters are not constrained. Using the original environment was really a dirty hack.

File: testsuite/tests/typing-misc/constraints.ml (hunk 2 above)
Comment (garrigue) [reply testsuite/tests/typing-misc/constraints.ml]: Having an exact error message is going to be difficult here. We could track the recursive use, and point it as a potential cause.

Comment (goldfirere) [reply testsuite/tests/typing-misc/constraints.ml]: I think that tracking the recursive use would be very helpful, if it's not too hard.

**Comment by garrigue** [comment]
Ok, new attempt.
I reverted to using the original (defective) environment, but modifying the warning so as to report this when the path is an identifier. This assumes that this is the only occurrence of a local identifier not appearing in the environment...
Should probably have a way to give this kind of information in the trace.

File: testsuite/tests/typing-missing-cmi/test.compilers.reference [hunk 3]
```diff
@@ -3,5 +3,7 @@ File "main.ml", line 1, characters 14-17:
                   ^^^
 Error: This expression has type "M.b" but an expression was expected of type
          "M.a"
-       "M.b" is abstract because no corresponding cmi file was found in path.
-       "M.a" is abstract because no corresponding cmi file was found in path.
+       Type "M.b" is abstract because no corresponding cmi file
+       was found in path.
+       Type "M.a" is abstract because no corresponding cmi file
+       was found in path.
```
Comment (gasche) [inline testsuite/tests/typing-missing-cmi/test.compilers.reference]: The line breaks here read like a regression caused by the use of `@ ` instead of ` ` in the error message. Why the change?

File: typing/printtyp.ml [hunk 4]
```diff
@@ -2439,9 +2439,16 @@ let warn_on_missing_def env ppf t =
...
+        match p with
+        | Path.Pident _ ->
+            fprintf ppf
+              "@,@[Type %a was considered abstract when checking constraints@ \
+               in this recursive type definition.@]"
+              (Style.as_inline_code path) p
+        | _ ->
+            fprintf ppf
+              "@,@[Type %a is abstract because no corresponding cmi file@ \
+               was found in path.@]" (Style.as_inline_code path) p
```
Comment (gasche) [inline typing/printtyp.ml]: It looks like you are catching *all errors on missing references*, and implicitly assuming that non-qualified paths all come from the specific errors of constraint checking, and specializing the error message for that. This sounds fragile and hard to understand and not a good idea.

**Review by gasche — COMMENTED** [review]
I don't like the new implementation. Can you explain why you decided against the previous approach? (Is it related to performance concerns on creating an environment?)

**Comment by garrigue** [comment]
This implementation is an answer to @goldfirere point that it would be nice to have a warning that we used the abstract version of the types. This seems to be the easiest way to do that. Performance is not an issue here.
I agree that this way of catching everything is not right. It should be easy to refine it by propagating the fact we are reporting an error coming from this precise point (e.g., add a flag to `Constraint_failed`).

I'm definitely not suggesting that we merge this code in its current state, I am rather exploring the design space.

**Comment by lpw25** [comment]
Doing this properly probably needs to put some kind of "reason" on the `Type_abstract` in the dummy type definition, and reporting that as part of failed expansions. I personally don't think that is worth doing for just this corner case -- which only experienced users or developers of the type system are likely to ever actually see. However, that would also more generally avoid giving the "missing cmi" error in the wrong place, something that has definitely happened before, and I think the required work could definitely be justified by that.

**Comment by garrigue** [comment]
@lpw25 That's precisely what I was trying to do yesterday, but I had no time to push then. Done now.
I think the code is reasonable, but this may still be an overkill.

File: testsuite/tests/typing-missing-cmi/test.compilers.reference (hunk 3 above)
Comment (garrigue) [reply testsuite/tests/typing-missing-cmi/test.compilers.reference]: Because the addition of `Type` at the beginning makes the line potentially too long for a standard terminal.
If you think there is no use for line breaks, we can also go back to ` `.

File: typing/printtyp.ml [hunk 5]
```diff
@@ -2432,19 +2432,20 @@ let explain mis ppf =
   | None -> ()
   | Some explain -> explain ppf
 
-let warn_on_missing_def env ppf t =
+type env_kind = Normal_env | Abstract_env of string
```
Comment (gasche) [inline typing/printtyp.ml]: Could you use a variant type for the `Abstract_env` payload, turned into a string only in the error-message production code?

File: typing/printtyp.ml [hunk 6]
```diff
@@ -2432,19 +2432,20 @@ let explain mis ppf =
...
   match get_desc t with
   | Tconstr (p,_,_) ->
     begin
       try
         ignore(Env.find_type p env : Types.type_declaration)
       with Not_found ->
-        match p with
-        | Path.Pident _ ->
+        match p, env_kind with
+        | Path.Pident _, Abstract_env where ->
```
Comment (gasche) [inline typing/printtyp.ml]: This match should have `env_kind` as the only scrutinee.

**Review by gasche — COMMENTED** [review]
(I didn't do a full review but just glanced at the last commit)

File: typing/printtyp.ml (hunk 5 above)
Comment (garrigue) [reply typing/printtyp.ml]: Is this a new policy ?

File: typing/printtyp.ml (hunk 6 above)
Comment (garrigue) [reply typing/printtyp.ml]: If this is not a `Pident`, the abstract environment is not the cause.
Of course, if we start to have other cases, this may not be the case, so in some way, your above suggestion of making the payload  a variant may make this code future-safe.

File: typing/printtyp.ml (hunk 5 above)
Comment (Octachron) [reply typing/printtyp.ml]: I don't know if this is a policy yet, but I concur that messages at destination to humans should be emitted in one piece if possible, in order to have the full context available.

I see three major benefits for this constraint on the error message:
* it can be read and copy edited in one place
* it can be formatted as a whole
* it can be translated with enough context to enable a good translation

**Comment by lpw25** [comment]
> @lpw25 That's precisely what I was trying to do yesterday, but I had no time to push then. Done now.

The code doesn't look like what I was suggesting. I was suggesting the following in `types.ml`:
```ocaml
type abstract_reason =
  | Missing_cmi
  | Recursive_dummy
  
and ('lbl, 'cstr) type_kind =
    Type_abstract of abstract_reason option
  | ...
```

**Comment by goldfirere** [comment]
This seems like a fine approach to me, and I’m happy with the final error message. Thanks, @garrigue! I will Approve, but others may want to chime in if they see a better way.

**Comment by garrigue** [comment]
> > @lpw25 That's precisely what I was trying to do yesterday, but I had no time to push then. Done now.
> 
> The code doesn't look like what I was suggesting. I was suggesting the following in `types.ml`: [...]

I see. This is indeed more elegant, but also more invasive.
This time we are lucky that looking at the path is enough to know the cause of abstraction, but your approach may be better if the problem shows up again.

**Comment by gasche** [comment]
Personally I still feel grumpy about the proposed approach. The problem is that even if we know that the environment we have is was intentionally built with some definitions missing, we can't know when we get a `Not_found` error whether this particular error comes from one of the definitions intentionally hidden from the environment, or from an actually undefined compilation unit somewhere else. So we do still need to rely on the criterion that "unqualified paths come from from our fake environment", which I find fragile and inelegant. (When I reviewed the change earlier I was under the impression that the match on `p, env_kind` was just a match on `env_kind` with the condition on `p` redundant, but in fact this is not the case.)

I see two ways to avoid relying on this criterion:

- Leo's proposal to mark abstract types with information, so that lookup of "types from the current recursive nest" return an abstract type marked as a dummy instead of failing with `Not_found` -- but I agree that it is more invasive.
- As a workaround, the `Partial_env_check_regularity` construct could come with a list/set of paths that are intentionally missing in the environment, and we could check membership into this list/set instead of relying on the `Pident _` catch-all.

Both approaches avoid the potential confusion that there is today between "Not_found errors on recursive neighbors" and "other Not_found errors".

**Comment by goldfirere** [comment]
Good point @gasche — this approach will get the message wrong if there
really is a missing cmi. I was too eager here!

**Comment by gasche** [comment]
So we discussed this during a type-system meeting and @garrigue was basically interested in trying out Leo's suggestions, with tweak to the definitions. (There is no `Missing_cmi` case as this one just raises `Not_found`.)

**Comment by garrigue** [comment]
Switched to Leo's approach. Of course this requires changes in many more files.

I kept the previous approach at https://github.com/cocti/ocaml/tree/flag_in_Constraint_failed .
Despite Gabriel's comment, I'm still convinced it is doing exactly what is intended (i.e. the reporting occurs exactly for the right paths), but I agree it is harder to review.

File: typing/typedecl.ml [hunk 7]
```diff
@@ -111,15 +111,19 @@ let enter_type rec_flag env sdecl (id, uid) =
   in
   let arity = List.length sdecl.ptype_params in
   if not needed then env else
+  let abstract_reason, type_manifest =
+    match sdecl.ptype_manifest, abstract_abbrevs with
+    | None, _             -> Abstract_def, None
+    | Some _, None        -> Abstract_def, Some (Btype.newgenvar ())
+    | Some _, Some reason -> reason, None
+  in
```
Comment (gasche) [inline typing/typedecl.ml:124]: This is obscure. Instead of taking an optional `abstract_abbrevs` parameter, could you take a non-optional `abstract_reason` parameter? There are two calls to `enter_type`, one with `Abstract_def` and one with `Abstract_rec_check_regularity`.

**Review by gasche — COMMENTED** [review]
I like the new approach better, thanks!

Reading the patch, I understand @lpw25's original proposal better. When he proposed to add a `Missing_cmi` reason, I think the idea was that the `Env.find_type*` lookup functions would return `Type_abstract Missing_cmi` *instead* of raising `Not_found`. This is a yet more invasive change but it would be fairly nice, because it could remove a lot of `Not_found` handling in the codebase (sharing the logic with the `Type_abstract _` case), and make it more robust in places where we don't currently catch `Not_found`.

This would also let us get rid of the `warn_on_missing_def*` machinery on printtyp: if each abstract type comes with an explanation of why it is abstract, we can print it (in non-obvious cases) when printing the explanation of the kind difference, instead of adding an extra explanation in an ad-hoc way as we do currently.

I think that these improvements could be implemented as separate / follow-up PRs, and that the current state is already an improvement. See minor comments on the code still.

File: typing/typedecl.ml (hunk 7 above)
Comment (garrigue) [reply typing/typedecl.ml:124]: My idea was that this is more versatile: one could eventually build abstract environments with other reasons.
But if you think this is an overgeneralization, I can go back to the original approach of using a boolean for `abstract_abbrevs`, and hard-coding the use of `Abstract_rec_check_regularity` here.
Or do you just mean that we should pass a value of type `abstract_reason option` at call sites ?

Comment (gasche) [reply typing/typedecl.ml:124]: After reading the code again, I understand the idea better, and I agree that my "simplification" proposed above was not such a good idea. Some documentation of the `abstract_abbrevs` parameter would have helped.

My understanding of the intended behavior: `enter_type` will turn datatype declarations into abstract types, and type synonyms into "dummy type synonym" of the form `type t = <newgenvar()>`. If `abstract_abbrevs` is provided, then all types (datatypes or synonyms) are made abstract with the provided reason.

Documenting this intention could be nice, because the code is not so clear. I also wonder what is the purpose of those 'dummy type synonyms" -- why not just turn type synonyms into abstract types as datatype declarations? (Ideally this could be explained in a comment.)

Comment (garrigue) [reply typing/typedecl.ml:124]: I added some explanations.

**Comment by garrigue** [comment]
> Reading the patch, I understand @lpw25's original proposal better. When he proposed to add a `Missing_cmi` reason, I think the idea was that the `Env.find_type*` lookup functions would return `Type_abstract Missing_cmi` _instead_ of raising `Not_found`. 

I thought of this too, but there is a problem: when we look up a type in the environment, we do not give its arity. So the dummy type returned in case of missing cmi would be nullary, requiring extra care when reporting. I suppose this may still be better than raising `Not_found`, but have to see how it fares. This is indeed not for this PR, but for a potential future one.

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
This PR was hit by a flaky CI failure: on AppVeyor (Windows) the test `memory-model/forbidden.byte` fails with (no output and) exit code -1073741819.
